### PR TITLE
Fixed canteen, made it use NBT aswell.

### PR DESCRIPTION
**Pull request by Dalethium**
Overall, basically this fixes a lot of issues with the canteen such as durability  ( #172 ) along with a few changes such as refill limits. The item uses NBT now ( #91 ) instead of item damage. You also no longer refill the canteen when it is full, and when you are underwater, if you can drink, you drink instead of refilling it, if you want to refill it you just shift -- this applies to all water sources too. There are 500 refills before it breaks so that it is fairly realistic (because this is what this mod is about, no?) The data is displayed via lore.

Everything works as far as I can see,  I did not change the code for filtering, or smelting, as I was not sure how you wanted it.  I have been using this code in a survival world with a few friends for days and it works fine. It will work if you update worlds, but the old canteens turn pink because the data method was changed. You can just make a new one though.

These changes were done on the web environment so there may be some syntax errors, but probably not. 

Side note: I am working on a temperature system re-work. It will be a lot more realistic. here is a snippet to show you how it will calculate temps.

[https://pastebin.com/L3JEdE36](url) pastebin link to snippit [it did not look pretty here]

### CODE REVIEW

**Comment by Dalethium** [comment]
the only thing that might not work is the canteen breaking, but honestly you can just remove the REFILL things, I set it to have 500 refills so that it is fairly realistic (because this is what this mod is about, no?) I have not tested the item breaking, that's it. everything else works as intended as far as I know

**Comment by Dalethium** [comment]
#91 <-- I was inspired to fix this by this

**Comment by Dalethium** [comment]
Fixed canteens being turned into glass bottles when used with cauldrons

**Comment by Dalethium** [comment]
sorry for the formatting in the last commit. Also, that last commit fixed a bug where the refill and use count would not show up on first refill. It also (probably) fixed a null pointer when you destroy the item. It has been tested, the item properly gets destroyed now.

**Comment by Dalethium** [comment]
Just so you know everything works now, I did not change the code for filtering, or smelting, as I was not sure how you wanted it. As for everything else, it all works. I have been using this code in a survival world with a few friends for days and it works fine. It will work if you update worlds, but the old canteens turn pink because the data method was changed. You can just make a new one though.

Side note: I am working on a temperature system re-work. It will be a lot more realistic. here is a snippet to show you how it will calculate temps.

`    public static float getBlockTemperature(IBlockState state) {
        Material material = state.getMaterial();
    	Fluid fluid = FluidRegistry.lookupFluidForBlock(state.getBlock());
        if (state.getBlock() == TANBlocks.campfire) {
        	if (state.getValue(BlockTANCampfire.BURNING) == true) {
        		return 1128.15F; //https://www.reference.com/sports-active-lifestyle/average-temperature-campfire-4833ec972e24a220#
        	} else {
        		return 0.0F; //http://ww2010.atmos.uiuc.edu/guides/maps/gifs/temp.scales.gif
        	}
        } else if (material == Material.FIRE) {
            return 1073.15F; //http://wildfiretoday.com/2011/02/26/at-what-temperature-does-a-forest-fire-burn/
        } else if (material == Material.LAVA) {
            return 1500.0F; //https://planetarygeomorphology.wordpress.com/2013/09/16/lava-temperatures-determine-flow-composition-on-earth-and-io/
        } else if (fluid != null){
        	return fluid.getTemperature(); //probably 300
        }
        
        return 0.0F;
    }
    `

**Comment by Dalethium** [comment]
Updated OP to make everything more clear.

**Comment by Adubbz** [comment]
There's a few things that concern me about this. First, we're not really looking to significantly change existing mechanics, canteens are very early game items and should remain as being 3 uses. Though TAN's mechanics do indeed have a basis in real life, that doesn't necessarily mean everything conforms to it 1:1. The thirst system does indeed have some flaws at this time, however I don't believe increasing the number of uses on canteens is the way to address it.

I'd also prefer not to rely on shift-clicking to refill canteens. The underwater issue can be resolved simply by preventing canteens from being refilled if they are already full, and instead going directly to the drinking code.

From the issues you've linked, the only real problem with the existing setup is the durability bar (which was working last I checked, although I something may have changed in an update that I didn't account for). Is it necessary to change to NBT to fix this? Also, one of the replies in the issues you've linked mentioned that this would cause problems with the durability bar, is that the case? Since that's a pretty important feature.

Also, though canteens are pretty cheap, I'm not too keen on making changes that would break people's existing worlds. It's something we can get away with more easily in BoP as it's simply the nature of the beast with biome/terrain-gen based mods, but even then, we haven't made a breaking change to BoP in quite some time. Unless there truly is no way of addressing the issues with the metadata system, I'd much prefer not to change it.

You also mentioned that you did this in the web environment. That's a pretty big red flag for me, since significant changes like this can only really be tested properly in an IDE. Additionally, there's a much lesser chance of there being syntax errors.

I haven't dived too deeply into the code itself yet, however at a quick glance, there appears to be missing override annotations for vanilla methods, copy-pasted method doc comments, and styling that clashes pretty significantly with the rest of our code. I'm generally not too fussy about styling, but when it's that drastic it doesn't really look good.

**Comment by Dalethium** [comment]
I have since tested it in game, and been using it for around a week with a few friends, the old ones still work, their texture is just broken - -  easily fixed by checking and setting the data value on refill. I can fix the formatting if need be.

**Comment by Dalethium** [comment]
Oh and it's still 3 uses, but you just can't refill it forever. The refills are changed, not the uses.

**Comment by Adubbz** [comment]
Oh okay, that's my bad, I misread what you wrote. I suppose it is an interesting enough mechanic, if you can get existing canteens to transfer over to it then there shouldn't be an issue on that front.

EDIT: Oops, wrong button

**Comment by Dalethium** [comment]
I fixed the formatting in the main changed class, hopefully that will help.

http://i.imgur.com/Rfn9AHC.png <-- a example of a frequently used canteen with my changes -- this is actually a screen shot of my survival server.

File: src/main/java/toughasnails/item/ItemCanteen.java [hunk 1]
```diff
@@ -30,162 +30,233 @@
...
                 WaterType waterType = getWaterType(stack);
-                
-                if (waterType == null) return 0.0F;
-                else return 1.0F;
+                if (waterType == null) {
+                    return 0.0F;
+                } else {
+                    return 1.0F;
+                }
             }
```
Comment (Adubbz) [inline src/main/java/toughasnails/item/ItemCanteen.java]: Unnecessary style changes

File: src/main/java/toughasnails/handler/thirst/FillBottleHandler.java [hunk 2]
```diff
@@ -58,7 +58,7 @@ public void onRightClickCauldron(PlayerInteractEvent.RightClickBlock event)
         EntityPlayer player = event.getEntityPlayer();
         IBlockState state = world.getBlockState(event.getPos());
         
-        if (state.getBlock() instanceof BlockCauldron && SyncedConfig.getBooleanValue(GameplayOption.ENABLE_THIRST))
+        if (player.getHeldItem(event.getHand()).getItem() == Items.GLASS_BOTTLE && state.getBlock() instanceof BlockCauldron && SyncedConfig.getBooleanValue(GameplayOption.ENABLE_THIRST))
         {
```
Comment (Adubbz) [inline src/main/java/toughasnails/handler/thirst/FillBottleHandler.java]: This seems to be unrelated to this change

File: src/main/java/toughasnails/item/ItemCanteen.java [hunk 3]
```diff
@@ -30,162 +30,233 @@
...
-        	return "item." + type.toString().toLowerCase() + "_water_canteen";
-        }
-        else
-        {
-        	return "item.empty_canteen";
+        if (type != null) {
+            return "item." + type.toString().toLowerCase() + "_water_canteen";
+        } else {
+            return "item.empty_canteen";
         }
```
Comment (Adubbz) [inline src/main/java/toughasnails/item/ItemCanteen.java]: Unnecessary style changes

File: src/main/java/toughasnails/item/ItemCanteen.java [hunk 4]
```diff
@@ -30,162 +30,233 @@
...
+    public int getMaxItemUseDuration(ItemStack stack) {
         return 32;
     }
 
     @Override
-    public EnumAction getItemUseAction(ItemStack stack)
-    {
+    public EnumAction getItemUseAction(ItemStack stack) {
         return EnumAction.DRINK;
     }
```
Comment (Adubbz) [inline src/main/java/toughasnails/item/ItemCanteen.java]: Unnecessary style changes

File: src/main/java/toughasnails/item/ItemCanteen.java [hunk 5]
```diff
@@ -30,162 +30,233 @@
...
-    {
-        return stack.getItemDamage() >> 2;
-    }
-    
     @Override
-    public int getMaxItemUseDuration(ItemStack stack)
-    {
+    public int getMaxItemUseDuration(ItemStack stack) {
         return 32;
     }
```
Comment (Adubbz) [inline src/main/java/toughasnails/item/ItemCanteen.java]: Unnecessary style changes

File: src/main/java/toughasnails/block/BlockRainCollector.java [hunk 6]
```diff
@@ -139,14 +139,9 @@ else if (item == TANItems.canteen)
...
-                    {
-                        heldItem.setItemDamage(3);
-                    }
-
-                    this.setWaterLevel(worldIn, pos, state, i - 1);
-                }
-            	
+            		return ((ItemCanteen) item).onItemRightClick(worldIn, playerIn, hand)
+            			.getType().equals(EnumActionResult.SUCCESS) ? true : false; 
+                }
```
Comment (Adubbz) [inline src/main/java/toughasnails/block/BlockRainCollector.java]: We use spaces, not tabs. Also the line wrapping isn't too great, better off keeping it as an editor setting, not putting it into the actual code

File: src/main/java/toughasnails/item/ItemCanteen.java [hunk 7]
```diff
@@ -30,162 +30,233 @@
 
 public class ItemCanteen extends Item
 {
-    public ItemCanteen()
-    {
-        this.addPropertyOverride(new ResourceLocation("filled"), new IItemPropertyGetter()
-        {
+    protected int REFILLS = 500;
```
Comment (Adubbz) [inline src/main/java/toughasnails/item/ItemCanteen.java]: Should be static final

File: src/main/java/toughasnails/item/ItemCanteen.java [hunk 8]
```diff
@@ -30,162 +30,233 @@
...
-                        stack.setItemDamage(1);
+                        comp.setInteger("water_type", 1);
+                        stack.setTagCompound(doRefill(setUses(comp, 3), stack, player)); 
+                        if (!player.capabilities.isCreativeMode) {
+                            cauldron.setWaterLevel(world, pos, state, level - 1);
+                            return true;
+                        }
+                        return true;
+                    }
+                } else if (state.getBlock() instanceof  BlockRainCollector) {
```
Comment (Adubbz) [inline src/main/java/toughasnails/item/ItemCanteen.java]: Double space inbetween instanceof and BlockRainCollector

**Review by Adubbz — CHANGES_REQUESTED** [review]
Few things to begin with, the existing style should be followed, not changed

File: src/main/java/toughasnails/handler/thirst/FillBottleHandler.java (hunk 2 above)
Comment (Dalethium) [reply src/main/java/toughasnails/handler/thirst/FillBottleHandler.java]: I am not sure if it was a bug with my code or not, but using a canteen on a cauldron would cause a watter bottle (not sure what kind, think it was filtered) to replace the canteen, this fixed it

Comment (Dalethium) [reply src/main/java/toughasnails/handler/thirst/FillBottleHandler.java]: not sure how to mark this as reviewed but it's as it should be

Comment (Adubbz) [reply src/main/java/toughasnails/handler/thirst/FillBottleHandler.java]: I guess it can be included with this, what I was saying though was that it's a separate fix that really should be split up from this

Comment (Dalethium) [reply src/main/java/toughasnails/handler/thirst/FillBottleHandler.java]: ooh ok. Whatever you think is best, seems easier to do it this way as it's already done and will be needed anyways.

**Comment by Dalethium** [comment]
@Adubbz did I miss anything?

**Comment by Adubbz** [comment]
Did you test what happens in terms of existing canteens under this new system? Otherwise we'll get a lot of 'my canteen doesn't work anymore' bug reports

**Comment by tyler489** [comment]
Might be able to make the old canteens have a lore that says this item is depreciated please put in crafting table to update . Alot of mods do so already.

**Comment by Forstride** [comment]
Closing this as Tough As Nails has been discontinued
